File: scala/binary-search-tree/src/main/scala/Bst.scala
import Bst.OrderedValues

case class Bst[T: OrderedValues](value: T,
    left: Option[Bst[T]] = None, right: Option[Bst[T]] = None)
{
  type Node = Option[Bst[T]]

  def insert(newValue: T): Bst[T] = {
    def insertNode(node: Node, insertF: Node => Bst[T]): Bst[T] = {
      val newNode: Node = node map (_.insert(newValue)) orElse Some(Bst(newValue))
      insertF(newNode)
    }

    if (newValue <= value) insertNode(left, n => copy(left = n))
    else insertNode(right, n => copy(right = n))
  }

  def foldLeft[A](z: A)(f: (A,T) => A): A = {
    val leftA = left.fold(z)(_.foldLeft(z)(f))
    val a = f(leftA, value)
    right.fold(a)(_.foldLeft(a)(f))
  }
}

object Bst {
  type OrderedValues[T] = T => Ordered[T]

  def fromList[T: Ordering](ts: List[T]): Bst[T] = {
    require(ts.nonEmpty, "list must be nonEmpty")
    ts.tail.foldLeft(Bst(ts.head))((bst, t) => bst insert t)
  }

  def toList[T](bst: Bst[T]): List[T] =
    bst.foldLeft(List.empty[T])((l,t) => l :+ t)
}

File: haskell/binary-search-tree/BST.hs
module BST (bstLeft, bstRight, bstValue, singleton, insert, fromList, toList) where

import Data.Maybe
import Control.Applicative ((<$>))

type BstValue = Int

type BstNode = Maybe Bst
data Bst = Bst { left :: BstNode, right :: BstNode, value :: BstValue }

bstLeft :: Bst -> BstNode
bstLeft = left

bstRight :: Bst -> BstNode
bstRight = right

bstValue :: Bst -> BstValue
bstValue = value

singleton :: BstValue -> Bst
singleton = Bst Nothing Nothing

insert :: BstValue -> Bst -> Bst
insert newValue (Bst l r v)
  | newValue <= v = insertLeft
  | otherwise = insertRight
  where
    insertLeft = Bst (insertInto l) r v
    insertRight = Bst l (insertInto r) v

    insertInto :: BstNode -> BstNode
    insertInto targetNode = Just $ fromMaybe newBst insertedNode
      where
        insertedNode :: BstNode
        insertedNode = insert newValue <$> targetNode
        newBst = singleton newValue

fromList :: [BstValue] -> Bst
fromList [] = error "fromList called on empty list"
fromList (x:xs) =
  foldl (flip insert) (singleton x) xs

toList :: Bst -> [BstValue]
toList (Bst l r v) = leftList ++ [v] ++ rightList
  where
    leftList = nodeList l
    rightList = nodeList r
    nodeList = maybe [] toList 
